addons/hr_contract/models/resource_calendar_leaves.py

```diff
@@ -0,0 +1,37 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from collections import defaultdict
+from datetime import datetime
+from pytz import timezone, utc
+
+from odoo import  models
+
+
+class ResourceCalendarLeaves(models.Model):
+    _inherit = 'resource.calendar.leaves'
+
+    def _compute_calendar_id(self):
+        Resource = self.env['resource.resource']
+        CalendarLeaves = self.env['resource.calendar.leaves']
+        leaves_by_resource = defaultdict(lambda: CalendarLeaves, {Resource: CalendarLeaves})
+        for leave in self:
+            leaves_by_resource[leave.resource_id] += leave
+        # pass leaves without resource_id to super
+        remaining = leaves_by_resource.pop(Resource)
+        for resource, leaves in leaves_by_resource.items():
+            contract = self.env['hr.contract'].search(
+                [('employee_id.resource_id', '=', resource.id), ('state', '=', 'open')],
+                limit=1,
+            )
+            if not contract:
+                remaining += leaves
+                continue
+            contract_start_dt = timezone(contract.resource_calendar_id.tz or 'UTC')\
+                .localize(datetime.fromisocalendar(*contract.date_start.isocalendar()))\
+                .astimezone(utc)\
+                .replace(tzinfo=None)
+            # don't modify leaves that fell under a previous contract
+            leaves.filtered(
+                lambda leave: contract_start_dt <= leave.date_from
+            ).calendar_id = contract.resource_calendar_id
```

Comment:
There can also be future planned leaves outside of the contract validity

Reply:
My thought process was that if a future contract changes `resource_id.calendar_id` again, it too will update the leave's calendar again, so I didn't consider filtering those out as crucial, but I could add it.